tools/codegen/model.py

```diff
@@ -210,52 +242,65 @@ def __post_init__(self) -> None:
-    out: Optional[NativeFunction]
+    out: NativeFunction
 
     def __post_init__(self) -> None:
-        test_sig: Optional[FunctionSchema] = None
+        test_sig: FunctionSchema = self.functional.func.signature()
         for f in self.functions():
-            if test_sig is None:
-                test_sig = f.func.signature()
-            else:
-                if test_sig != f.func.signature():
-                    raise AssertionError(
-                        "NativeFunctionGroup constructed from two NativeFunctions "
-                        f"that don't have matching signatures: {test_sig} != {f.func.signature()}"
-                    )
+            if test_sig != f.func.signature():
+                raise AssertionError(
+                    "StructuredNativeFunctions constructed from two NativeFunctions "
+                    f"that don't have matching signatures: {test_sig} != {f.func.signature()}"
+                )
+        assert self.functional.func.kind() == SchemaKind.functional
+        assert self.functional.structured_delegate == self.out.func.name, \
+            f"{self.functional.func.name} delegates to {self.functional.structured_delegate} " \
+            f"but its actual delegate is {self.out.func.name}"
+        assert self.out.func.kind() == SchemaKind.out
+        assert self.out.structured
+        # For now, structured composite kernels are not supported (need some
+        # design work to figure out how to make the composite case work)
+        assert self.out.dispatch.keys() != {'Math'}
+        if self.inplace is not None:
+            assert self.inplace.func.kind() == SchemaKind.inplace
+            assert self.inplace.structured_delegate == self.out.func.name
 
     def signature(self) -> 'FunctionSchema':
-        if self.out is not None:
-            return self.out.func.signature()
-        elif self.functional is not None:
-            return self.functional.func.signature()
-        elif self.inplace is not None:
-            return self.inplace.func.signature()
-        else:
-            raise AssertionError("invalid NativeFunctionGroup has no NativeFunctions")
+        return self.out.func.signature()
 
     def functions(self) -> Iterator[NativeFunction]:
-        if self.out is not None:
-            yield self.out
-        if self.functional is not None:
-            yield self.functional
+        yield self.out
+        yield self.functional
         if self.inplace is not None:
             yield self.inplace
 
     @staticmethod
-    def from_dict(d: Dict[SchemaKind, NativeFunction]) -> 'NativeFunctionGroup':
+    def from_dict(d: Dict[SchemaKind, NativeFunction]) -> Optional['StructuredNativeFunctions']:
         functional = d.get(SchemaKind.functional)
         inplace = d.get(SchemaKind.inplace)
         out = d.get(SchemaKind.out)
-        return NativeFunctionGroup(
+        if functional is None or out is None or not out.structured:
+            return None
```

Comment:
Should it be an error to specify out as structured without declaring a functional variant? (Or an assert if it's checked upstream, I guess)

Don't know if there are callers that need this to be permissive, but e.g. the code path that runs through `flatten_pre_group()` looks like it might benefit from error checking.

Reply:
It's checked separately, in `validate_unstructured`. I didn't put error tests here because I don't have error context available.